src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs

```diff
@@ -283,23 +318,48 @@ async Task<(Solution Solution, bool IsForked)> GetLspSolutionForWorkspaceAsync(W
                 return (workspaceCurrentSolution, IsForked: false);
             }
 
-            // Step 3: See if we can reuse a previously forked solution.
+            // Step 4: See if we can reuse a previously forked solution.
             if (cachedSolution != default && cachedSolution.forkedFromVersion == workspaceCurrentSolution.WorkspaceVersion)
-            {
                 return (cachedSolution.solution, IsForked: true);
-            }
 
-            // Step 4: Fork a new solution from the workspace with the LSP text applied.
+            // Step 5: Fork a new solution from the workspace with the LSP text applied.
             var lspSolution = workspaceCurrentSolution;
             foreach (var (uri, workspaceDocuments) in documentsInWorkspace)
-            {
                 lspSolution = lspSolution.WithDocumentText(workspaceDocuments.Select(d => d.Id), _trackedDocuments[uri]);
-            }
 
             // Remember this forked solution and the workspace version it was forked from.
             _cachedLspSolutions[workspace] = (workspaceCurrentSolution.WorkspaceVersion, lspSolution);
             return (lspSolution, IsForked: true);
         }
+
+        void OpenOrEditDocumentsInMutatingWorkspace(Workspace workspace)
+        {
+            if (workspace is IMutatingLspWorkspace)
+            {
+                var currentSolution = workspace.CurrentSolution;
+                foreach (var (uri, sourceText) in _trackedDocuments)
+                {
+                    foreach (var documentId in currentSolution.GetDocumentIds(uri))
+                    {
+                        // If not already open in this workspace, open it.  If already opened, update it to this latest text value.
+                        if (workspace.IsDocumentOpen(documentId))
+                        {
+                            // Only bother updating if we 
+                            var document = currentSolution.GetDocument(documentId);
+                            if (document != null && document.TryGetText(out var existingText) && existingText == sourceText)
+                                continue;
+
+                            workspace.OnDocumentTextChanged(documentId, sourceText, PreservationMode.PreserveIdentity);
+                        }
+                        else
+                        {
+                            // TODO(cyrusn): What is the right value for isCurrentContext here?
+                            workspace.OnDocumentOpened(documentId, sourceText.Container, isCurrentContext: false);
```

Comment:
Ah I think this is for linked documents.  I don't think at this point we know or have a project context for the document.  It seems fine like this (to open both docs with neither as the current context) - but definitely need a test that includes linked docs.